clang/test/SemaCXX/cheri/overloaded-function-template-arg.cpp

```diff
@@ -34,22 +34,32 @@ void instantiate() {
 
 #if __has_feature(capabilities)
 template <class T, T *__capability Func>
+// hybrid-note@-1 3 {{template parameter is declared here}}
 struct c_cap {
   static constexpr T *__capability function = Func;
 };
 template <class T, class Arg>
 c_cap<T, a> test_overloaded_fn_cap(Arg arg) {
   // expected-note@-1{{candidate template ignored: substitution failure [with T = void (char)]: address of overloaded function 'a' does not match required type 'void (char)'}}
+  // hybrid-note@-2{{candidate template ignored: substitution failure [with T = void (int)]: non-type template argument of type 'void (int)' cannot be converted to a value of type 'void (* __capability)(int)'}}
+  // hybrid-note@-3{{candidate template ignored: substitution failure [with T = void (long)]: non-type template argument of type 'void (long)' cannot be converted to a value of type 'void (* __capability)(long)'}}
   c_cap<T, a>::function(arg);
   return {};
 }
 void instantiate_cap() {
-  // The following (valid) expressions were previously rejected by CHERI Clang:
+  // TODO: These were previously allowed, but had the wrong semantics, since
+  // T * __capability dropped the __capability when instantiated. Should these
```

Comment:
It would be more consistent if implicit also allowed the conversion from `void (int)` -> `void (* __capability)(int)` rather than only decaying to `void (*)(int)`. However, I would only bother to implement that that if it's trivial. Forcing casts for these rare cases seems perfectly reasonable and I think we should avoid implicit conversions wherever possible.

Reply:
As long as it works fine in purecap, I don't think it matters what hybrid does for that code.